src/daemon/utils_ssl.c

```diff
@@ -0,0 +1,104 @@
+/**
+ * collectd - src/utils_ssl.c
+ * Copyright (C) 2007-2010  Florian octo Forster
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; only version 2 of the License is applicable.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Author:
+ *   Florian octo Forster <octo at collectd.org>
+ *   Toni Moreno <toni at gmail.com>
+ */
+
+ /* Code taken from The example code from Jeremy Brown:
+ * http://curl.haxx.se/libcurl/c/opensslthreadlock.html
+ **/
+
+#if HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#ifndef HAVE_OPENSSL
+int cd_ssl_multithread_setup(void)
+{
+	return 0;
+}
+
+int cd_ssl_multithread_cleanup(void)
+{
+	return 0;
+}
+#else /*DEFINED HAVE_OPENSSL*/
+
+
+#include <stdio.h>
+#include <pthread.h>
+#include <openssl/err.h>
+#include <openssl/crypto.h>
+
+#define MUTEX_TYPE       pthread_mutex_t
+#define MUTEX_SETUP(x)   pthread_mutex_init(&(x), NULL)
+#define MUTEX_CLEANUP(x) pthread_mutex_destroy(&(x))
+#define MUTEX_LOCK(x)    pthread_mutex_lock(&(x))
+#define MUTEX_UNLOCK(x)  pthread_mutex_unlock(&(x))
+#define THREAD_ID        pthread_self(  )
+
+
+/* This array will store all of the mutexes available to OpenSSL. */
+static MUTEX_TYPE *mutex_buf= NULL;
+
+
+static void locking_function(int mode, int n, const char * file, int line)
+{
+  if (mode & CRYPTO_LOCK)
+    MUTEX_LOCK(mutex_buf[n]);
+  else
+    MUTEX_UNLOCK(mutex_buf[n]);
+}
+
+static unsigned long id_function(void)
+{
+  return ((unsigned long)THREAD_ID);
+}
+
+int cd_ssl_multithread_setup(void)
+{
+  int i;
+  /*check if already initialized */
+  if(mutex_buf != NULL) return 1;
+  /*if not begin callback initialization*/
+  mutex_buf = malloc(CRYPTO_num_locks(  ) * sizeof(MUTEX_TYPE));
+  if (!mutex_buf)
+    return 0;
+  for (i = 0;  i < CRYPTO_num_locks(  );  i++)
+    MUTEX_SETUP(mutex_buf[i]);
+  CRYPTO_set_id_callback(id_function);
+  CRYPTO_set_locking_callback(locking_function);
+  return 1;
+}
+
+int cd_ssl_multithread_cleanup(void)
```

Comment:
I've been thinking, and you are right  that should be properly defined where to call these two functions:

cd_ssl_multithread_setup()
cd_ssl_multithread_cleanup()

Once defined that should be always called on "init" and "shutdown"   plugin callbacks. 

It is strictly necessary to add more complexity to that? I can not imagine some situation where these two functions should be called in a multithread environment  ( in plugin_read_cb by example ?).

@tokkee  if you think it is absolutely necessary I will do, but I don't think so. 

I will wait  your answer before any other change. 

Thank you !!